rfd/0214-identifier-first-login.md

```diff
@@ -0,0 +1,193 @@
+
+```yaml
+kind: saml
+metadata:
+  name: testconnector
+  revision: 60443273-15b2-4e5b-a2e2-9338439eac37
+spec:
+  acs: https://<cluster-url>/v1/webapi/saml/acs/testconnector
+  attributes_to_roles:
+    - name: groups
+      roles:
+        - editor
+      value: okta-admin
+    - name: groups
+      roles:
+        - auditor
+      value: okta-auditor
+    - name: groups
+      roles:
+        - access
+      value: okta-dev
+  audience: https://<cluster-url>/v1/webapi/saml/acs/testconnector
+  user_matchers: ["^joe@foo.com$", ".*@goteleport\.com$"]
+  cert: ''
+  display: Okta
+  entity_descriptor: |
+    <?xml version="1.0" encoding="UTF-8"?>
+    <md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="http://www.example.com/00000000000000000000">
+    ...
+    </md:EntityDescriptor>
+  entity_descriptor_url: ''
+  issuer: http://www.example.com/00000000000000000000
+  service_provider_issuer: https://teleport.com/v1/webapi/saml/acs/testconnector
+  signing_key_pair:
+    cert: |
+      -----BEGIN CERTIFICATE-----
+      ...
+      -----END CERTIFICATE-----
+    private_key: |
+      -----BEGIN RSA PRIVATE KEY-----
+      ...
+      -----END RSA PRIVATE KEY-----
+  sso: https://www.example.com/app/00000000000000000000/sso/saml
+version: v2
+```
+
+### Implementation
+
+Upon loading the Web UI, the server returns a web config which contains some details about the cluster
+including the Teleport edition, login methods and auth connectors (stored in `window['GRV_CONFIG']`).
+This will now also include the RegEx matchers per auth connector.
+
+All matching and filtering will be done on the client side. This is so that we can avoid having to
+introduce any additional calls to the backend once the user has provided a username. This is faster
+and more performant, and also reduces complexity since it means we won't need to add and maintain any new
+endpoints. Remembered usernames will be kept in `localStorage`.
+
+#### Proto changes
+
+`SAMLConnectorSpecV2`, `OIDCConnectorSpecV3`, and `GithubConnectorSpecV3` will need to be updated to include the `user_matchers` field.
+
+```proto
+// SAMLConnectorSpecV2 is a SAML connector specification.
+message SAMLConnectorSpecV2 {
+  ...
+  // UserMatchers is a set of RegEx patterns to narrow down which username(s) this auth connector should
+  // match for identifier-first login.
+  repeated string UserMatchers = 19 [(gogoproto.jsontag) = "user_matchers,omitempty"]
+}
+```
+
+```proto
+// OIDCConnectorSpecV3 is an OIDC connector specification.
+//
+// It specifies configuration for Open ID Connect compatible external
+// identity provider: https://openid.net/specs/openid-connect-core-1_0.html
+message OIDCConnectorSpecV3 {
+ ...
+  // UserMatchers is a set of RegEx patterns to narrow down which username(s) this auth connector should
+  // match for identifier-first login.
+  repeated string UserMatchers = 21 [(gogoproto.jsontag) = "user_matchers,omitempty"]
```

Comment:
For OIDC we have an optional `username_claim` property. Will the `user_matchers` match against whatever claim we specify there, or will they always match against email?

Reply:
It will always match against the email. The OIDC `username_claim` relies on a claim returned by the IdP, but the matching for identifier-first login is completely detached from that flow and doesn't actually interact with the IdP or IdP response directly. The `user_matchers` is only used to detect whether a user-inputted string (which can be anything) should be matched to the auth connector, but this all happens on the login page before any SSO flow is initiated.